File: gr00t/Isaac-GR00T/examples/H20/robots/sdk_package/example/h10-w/low_level/lrcu_cmd.cpp
// lrcu_cmd.cpp

#include <iostream>
#include <thread>
#include <chrono>

#include "common/common_def.h"
#include "common/h10w_def.h"
#include "h10-w/h10w_low_level.h"

using namespace humanoid;
using namespace humanoid::h10w;

int main() {
  std::cout << "=== LRCU Command Test ===" << std::endl;

  H10wLowLevel low_level;
  LRCUCmd lrcu_cmd;

  // Test LED control for 10 seconds
  constexpr int DURATION_MS = 10000;
  constexpr int CMD_INTERVAL_MS = 5000;
  constexpr int cmd_count = DURATION_MS / CMD_INTERVAL_MS;

  for (int i = 0; i < cmd_count; ++i) {
    auto start = std::chrono::high_resolution_clock::now();

    // Set LED parameters
    lrcu_cmd.mode = 1;  // Enable LED

    // Cycle through different colors
    int color_phase = i % 6;
    switch (color_phase) {
      case 0:  // Red
        lrcu_cmd.led_r = 255;
        lrcu_cmd.led_g = 0;
        lrcu_cmd.led_b = 0;
        lrcu_cmd.led_mode = 1;  // Solid color
        lrcu_cmd.led_frequency = 0;
        std::cout << "Setting LED to RED" << std::endl;
        break;
      case 1:  // Green
        lrcu_cmd.led_r = 0;
        lrcu_cmd.led_g = 255;
        lrcu_cmd.led_b = 0;
        lrcu_cmd.led_mode = 1;
        lrcu_cmd.led_frequency = 0;
        std::cout << "Setting LED to GREEN" << std::endl;
        break;
      case 2:  // Blue
        lrcu_cmd.led_r = 0;
        lrcu_cmd.led_g = 0;
        lrcu_cmd.led_b = 255;
        lrcu_cmd.led_mode = 1;
        lrcu_cmd.led_frequency = 0;
        std::cout << "Setting LED to BLUE" << std::endl;
        break;
      case 3:  // Yellow
        lrcu_cmd.led_r = 255;
        lrcu_cmd.led_g = 255;
        lrcu_cmd.led_b = 0;
        lrcu_cmd.led_mode = 2;  // Blinking
        lrcu_cmd.led_frequency = 1;
        std::cout << "Setting LED to YELLOW (Blinking)" << std::endl;
        break;
      case 4:  // Cyan
        lrcu_cmd.led_r = 0;
        lrcu_cmd.led_g = 255;
        lrcu_cmd.led_b = 255;
        lrcu_cmd.led_mode = 3;  // Breathing
        lrcu_cmd.led_frequency = 5;
        std::cout << "Setting LED to CYAN (Breathing)" << std::endl;
        break;
      case 5:  // Magenta
        lrcu_cmd.led_r = 255;
        lrcu_cmd.led_g = 0;
        lrcu_cmd.led_b = 255;
        lrcu_cmd.led_mode = 2;  // Blink
        lrcu_cmd.led_frequency = 2;
        std::cout << "Setting LED to MAGENTA (Blink)" << std::endl;
        break;
    }

    // Send LRCU command
    int ret = low_level.writeLrcuCmd(lrcu_cmd);
    if (ret != 0) {
      std::cout << "Failed to write LRCU command!" << std::endl;
      return -1;
    }

    auto end = std::chrono::high_resolution_clock::now();
    auto elapsed =
        std::chrono::duration_cast<std::chrono::milliseconds>(end - start);
    std::this_thread::sleep_for(std::chrono::milliseconds(CMD_INTERVAL_MS) -
                                elapsed);
  }

  // Turn off LED at the end
  lrcu_cmd.mode = 0;
  low_level.writeLrcuCmd(lrcu_cmd);
  std::cout << "LED turned off" << std::endl;

  std::cout << "LRCU command test completed successfully" << std::endl;
  return 0;
}

File: gr00t/Isaac-GR00T/examples/H20/robots/sdk_package/example_py/h10-w/low_level/lrcu_cmd.py
#!/usr/bin/env python3

import humanoid_sdk_py
from humanoid_sdk_py import h10w
import time


def main():
    print("=== LRCU Command Test ===")

    # Initialize low level object
    low_level = h10w.H10wLowLevel()

    # Test LED control for 10 seconds
    DURATION_MS = 10000
    CMD_INTERVAL_MS = 500
    cmd_count = DURATION_MS // CMD_INTERVAL_MS

    for i in range(cmd_count):
        start_time = time.time()

        # Create LRCU command
        lrcu_cmd = h10w.LRCUCmd()

        # Set LED parameters
        lrcu_cmd.mode = 1  # Enable LED

        # Cycle through different colors
        color_phase = i % 6
        if color_phase == 0:  # Red
            lrcu_cmd.led_r = 255
            lrcu_cmd.led_g = 0
            lrcu_cmd.led_b = 0
            lrcu_cmd.led_mode = 1  # Solid color
            print("Setting LED to RED")
        elif color_phase == 1:  # Green
            lrcu_cmd.led_r = 0
            lrcu_cmd.led_g = 255
            lrcu_cmd.led_b = 0
            lrcu_cmd.led_mode = 1
            print("Setting LED to GREEN")
        elif color_phase == 2:  # Blue
            lrcu_cmd.led_r = 0
            lrcu_cmd.led_g = 0
            lrcu_cmd.led_b = 255
            lrcu_cmd.led_mode = 1
            print("Setting LED to BLUE")
        elif color_phase == 3:  # Yellow
            lrcu_cmd.led_r = 255
            lrcu_cmd.led_g = 255
            lrcu_cmd.led_b = 0
            lrcu_cmd.led_mode = 2  # Blinking
            lrcu_cmd.led_frequency = 2
            print("Setting LED to YELLOW (Blinking)")
        elif color_phase == 4:  # Cyan
            lrcu_cmd.led_r = 0
            lrcu_cmd.led_g = 255
            lrcu_cmd.led_b = 255
            lrcu_cmd.led_mode = 3  # Breathing
            lrcu_cmd.led_frequency = 1
            print("Setting LED to CYAN (Breathing)")
        elif color_phase == 5:  # Magenta
            lrcu_cmd.led_r = 255
            lrcu_cmd.led_g = 0
            lrcu_cmd.led_b = 255
            lrcu_cmd.led_mode = 2  # Blink
            lrcu_cmd.led_frequency = 3
            print("Setting LED to MAGENTA (Blink)")

        # Send LRCU command
        ret = low_level.writeLrcuCmd(lrcu_cmd)
        if ret != 0:
            print("Failed to write LRCU command!")
            return -1

        elapsed_time = (time.time() - start_time) * 1000
        sleep_time = CMD_INTERVAL_MS - elapsed_time
        if sleep_time > 0:
            time.sleep(sleep_time / 1000.0)

    # Turn off LED at the end
    lrcu_cmd.mode = 0
    low_level.writeLrcuCmd(lrcu_cmd)
    print("LED turned off")

    print("LRCU command test completed successfully")
    return 0


if __name__ == "__main__":
    exit(main())
